end2end-test-examples/spanner/src/main/java/io/grpc/spanner/TestMain.java

```diff
@@ -126,84 +138,169 @@ public static void main(String[] args) throws Exception {
           String.format("%.4f", (numOtherErr / numTotalRPC))
         });
 
+    tcpkillManageThread.isExit = true;
+    iptablesManageThread.isExit = true;
     spannerClient.cleanUp();
   }
 
   /*
-   * A side thread used to do tcpkill or iptables
+   * This thread manages iptables operation
+   */
+  private static class IptablesManageThread extends Thread {
+    public volatile boolean isExit = false;
+    private int iptablesMs;
+    private int iptablesDurationMs;
+    private String logFilename;
+
+    IptablesManageThread(String name, int interval, int duration) {
+      logFilename = name;
+      iptablesMs = interval;
+      iptablesDurationMs = duration;
+    }
+
+    @Override
+    public void run() {
+      logger.log(Level.FINE, "Iptables manage thread: start, log file is {0}", logFilename);
+      while (!isExit) {
+        IptablesThread iptablesThread = new IptablesThread(logFilename, iptablesDurationMs);
+        iptablesThread.start();
+        try {
+          Thread.sleep(iptablesMs + iptablesDurationMs);
+        } catch (InterruptedException e) {
+        }
+      }
+      logger.log(Level.FINE, "Iptables manage thread: exit.");
+    }
+  }
+
+  /*
+   * This thread use iptables to blackhole a port
+   */
+  private static class IptablesThread extends Thread {
+    private String logFilename;
+    private int iptablesDurationMs;
+
+    IptablesThread(String name, int duration) {
+      logFilename = name;
+      iptablesDurationMs = Math.max(duration, 5000);
+    }
+
+    @Override
+    public void run() {
+      logger.log(Level.FINE, "Iptables thread: start");
+      if (logFilename.isEmpty()) {
+        logger.log(Level.FINE, "Please specify logFilename and enable fineLog");
+        return;
+      }
+      try {
+        String grep_cmd =
+            "grep 'OUTBOUND HEADERS' "
+                + logFilename
+                + " | grep 'ExecuteStreamingSql' | tail -1 | awk '{print $4}' | awk '{print"
+                + " substr($0,length($0)-4)}' | tr '\n' ' '";
+        Process grep_process = Runtime.getRuntime().exec(new String[] {"/bin/sh", "-c", grep_cmd});
+        grep_process.waitFor();
+        String port =
+            CharStreams.toString(new InputStreamReader(grep_process.getInputStream(), "UTF-8"));
+        if (port.isEmpty()) {
+          logger.log(Level.INFO, "Iptables thread: could not find a valid port");
+          return;
+        }
+
+        String blackhole_cmd = "sudo iptables -I INPUT -p tcp --dport " + port + " -j DROP";
+        logger.log(Level.FINE, "Iptables thread: blockhole port " + port);
+        Process blackhole_process =
+            Runtime.getRuntime().exec(new String[] {"/bin/sh", "-c", blackhole_cmd});
+        blackhole_process.waitFor();
+
+        Thread.sleep(iptablesDurationMs);
+
+        String whitelist_cmd = "sudo iptables -I INPUT -p tcp --dport " + port + " -j ACCEPT";
```

Comment:
AFAIK the rules in iptables gets processed from the top to the bottom, and if we add accept rule below the drop rule it won't work because the processing stops on the drop rule above. If this is true we can either always insert all rules to the top or flush the rules for the input chain instead of adding accept rule (but this may lead to removing some rules not created by this program)

Reply:
`-I` means insert, without specifying the position the default is the beginning. So we add new rules at the top, and `ACCEPT` can work.

Reply:
Oh... I missed that, I thought it was append.